### PR TITLE
[IMP] Allow to define a custom relation table on custom many2many fields

### PR DESCRIPTION
**Pull request by ghost**
Define the columns and name of the relation table of custom many2many fields
Improve reflection of existing many2many fields (@rco-odoo)
Add onchange for many2many fields (@rco-odoo)

### CODE REVIEW

**Comment by tgr-odoo** [comment]
@rco-odoo

**Comment by rco-odoo** [comment]
Functionally it works :+1: 

However, if a many2many field was created with default names, you cannot see the chosen table and column names :disappointed: Therefore introducing an "inverse" many2many is almost impossible if you don't know how names are chosen :wrench:

**Comment by rco-odoo** [comment]
@sve-odoo @ahu-odoo Can you test the branch, please?

**Comment by rco-odoo** [comment]
Note: because of the change in `base.sql`, the registry will not be able to load on an existing database.

Here is how to migrate existing databases:
 - update the DB schema with:

        ALTER TABLE ir_model_fields ADD COLUMN relation_table varchar;
        ALTER TABLE ir_model_fields ADD COLUMN column1 varchar;
        ALTER TABLE ir_model_fields ADD COLUMN column2 varchar;

 - upgrade all modules, or write an equivalent script to fill in the given columns on many2many fields.

**Comment by rco-odoo** [comment]
@KangOl can you have a look at this? Thanks!

File: openerp/addons/base/ir/ir_model.py [hunk 1]
```diff
@@ -287,26 +290,68 @@ def _size_gt_zero_msg(self, cr, user, ids, context=None):
         ('size_gt_zero', 'CHECK (size>=0)',_size_gt_zero_msg ),
     ]
 
+    @api.onchange('ttype', 'model_id', 'relation')
+    def _onchange_ttype(self):
+        if self.ttype == 'many2many' and self.model_id and self.relation:
+            rel1 = self.relation.replace('.', '_')
+            rel2 = self.model_id.model.replace('.', '_')
+            self.relation_table = 'x_%s_%s_%s_rel' % (rel1, rel2, self.name)
```
Comment (KangOl) [inline openerp/addons/base/ir/ir_model.py]: should sort (rel1, rel2)

Comment (rco-odoo) [inline openerp/addons/base/ir/ir_model.py]: Okay, but then filling in values for existing fields MUST be done explicitly by migration!

Comment (rco-odoo) [inline openerp/addons/base/ir/ir_model.py]: @KangOl should I generate a table name like `openerp.osv.fields.many2many._sql_names`?

    model1 = self.env[self.relation]
    model2 = self.env[self.model_id.model]
    self.relation_table = "x_%s_%s_rel" % tuple(sorted([model1._table, model2._table]))
    self.column1 = "%s_id" % model1._table
    self.column2 = "%s_id" % model2._table

Comment (KangOl) [inline openerp/addons/base/ir/ir_model.py]: Yes, better

**Pull request merged**